src/main/java/jp/co/soramitsu/sora/crypto/Crypto.java

```diff
@@ -0,0 +1,145 @@
+package jp.co.soramitsu.sora.crypto;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.security.KeyPair;
+import java.security.PublicKey;
+import java.time.Instant;
+import java.util.LinkedList;
+import java.util.List;
+import javax.validation.constraints.NotNull;
+import jp.co.soramitsu.sora.crypto.algorithms.RawSignatureStrategy;
+import jp.co.soramitsu.sora.crypto.algorithms.RawSignatureStrategy.SignatureSuiteException;
+import jp.co.soramitsu.sora.crypto.algorithms.SignatureSuiteRegistry;
+import jp.co.soramitsu.sora.crypto.hash.RawDigestStrategy;
+import jp.co.soramitsu.sora.util.Bencoder;
+import lombok.Getter;
+import lombok.Setter;
+
+
+public class Crypto {
+
+  @Setter
+  @Getter
+  private RawDigestStrategy digestStrategy;
+
+  public Crypto(@NotNull RawDigestStrategy digestStrategy) {
+    this.digestStrategy = digestStrategy;
+  }
+
+  private byte[] createVerifyHash(VerifiableJson document, ProofProxy proof)
+      throws CreateVerifyHashException {
+    ByteArrayOutputStream os = new ByteArrayOutputStream();
+    Bencoder bencoder = new Bencoder(UTF_8, os);
+
+    // sanitize inputs
+    this.sanitizeDocument(document);
+    this.sanitizeProof(proof);
+
+    // encode input into a string
+    try {
+      bencoder.encode(document.serializeAsMap());
+      bencoder.encode(proof.serializeAsMap());
+    } catch (IOException e) {
+      throw new CreateVerifyHashException(e);
+    }
+
+    return digestStrategy.digest(os.toByteArray());
+  }
+
+  private void sanitizeDocument(VerifiableJson document) {
+    document.setProof(null);
+  }
+
+  private void sanitizeProof(ProofProxy proof) {
+    // if "created" is empty, then set current time as "created"
+    if (proof.getCreated() == null) {
+      proof.setCreated(Instant.now());
+    }
+
+    proof.setSignatureValue(null);
+  }
+
+  private RawSignatureStrategy getSuite(String type) throws InvalidAlgorithmException {
+    // find appropriate digital signature algorithm
+    if (!SignatureSuiteRegistry.has(type)) {
+      throw new InvalidAlgorithmException(type + " signature suite is not implemented");
+    }
+
+    return SignatureSuiteRegistry.get(type);
+  }
+
+  public void sign(VerifiableJson document, KeyPair keypair, ProofProxy proof)
+      throws InvalidAlgorithmException, CreateVerifyHashException, SignatureSuiteException {
+    RawSignatureStrategy signer = getSuite(proof.getType());
+
+    // backup proofs
+    List<ProofProxy> proofs = document.getProof();
+
+    // hash and sign input
+    byte[] hash = createVerifyHash(document, proof);
+    byte[] signature = signer.rawSign(hash, keypair);
+
+    // include signature into a proof, add it to saved proofs
+    proof.setSignatureValue(signature);
+    if (proofs == null) {
+      proofs = new LinkedList<>();
```

Comment:
why did you choose LinkedList?

Reply:
it does not matter mostly, I need only single add operation. Is there a better candidate?

Reply:
yes, use arraylist